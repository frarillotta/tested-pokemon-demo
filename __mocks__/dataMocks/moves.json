[
	{
	   "move":{
		  "name":"mega-punch",
		  "url":"https://pokeapi.co/api/v2/move/5/"
	   },
	   "version_group_details":[
		  {
				 "level_learned_at":0,
				 "move_learn_method":{
				"name":"machine",
				"url":"https://pokeapi.co/api/v2/move-learn-method/4/"
				 },
				 "version_group":{
				"name":"red-blue",
				"url":"https://pokeapi.co/api/v2/version-group/1/"
				 }
		  }
	   ]
	},
	{
	   "move":{
		  "name":"ice-punch",
		  "url":"https://pokeapi.co/api/v2/move/8/"
	   },
	   "version_group_details":[
		  {
				 "level_learned_at":0,
				 "move_learn_method":{
				"name":"machine",
				"url":"https://pokeapi.co/api/v2/move-learn-method/4/"
				 },
				 "version_group":{
				"name":"gold-silver",
				"url":"https://pokeapi.co/api/v2/version-group/3/"
				 }
		  },
		  {
				 "level_learned_at":0,
				 "move_learn_method":{
				"name":"machine",
				"url":"https://pokeapi.co/api/v2/move-learn-method/4/"
				 },
				 "version_group":{
				"name":"crystal",
				"url":"https://pokeapi.co/api/v2/version-group/4/"
				 }
		  }
	   ]
	  }
 ]