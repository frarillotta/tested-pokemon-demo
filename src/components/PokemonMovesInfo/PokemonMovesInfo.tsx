import type { Move } from "../../types";
import AutoSizer from "react-virtualized-auto-sizer";
import { FixedSizeList as List } from 'react-window';
import styled from "styled-components";
import FormControl from '@material-ui/core/FormControl';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';
import InputLabel from '@material-ui/core/InputLabel';
import { useState, useMemo } from "react";
import { capitalizeFirstLetter } from "../../utils";

type PokemonMovesInfoType = {
	moves: Move[],
	className?: string
};

const PokemonMovesInfo = ({moves, className = ""}: PokemonMovesInfoType) => {
	
	const availableGames = useMemo<string[]>(() => moves.reduce<string[]>((accumulator, currentVal) => {
		currentVal.version_group_details.forEach((groupDetails)=>{
			const name = groupDetails.version_group.name
			if (!accumulator.includes(name)) accumulator.push(name)
		});
		return accumulator;
	}, []), []);
	
	const [currentSelection, setSelection] = useState(availableGames[0]);
	const handleChange = (event) => {
		setSelection(event.target.value);
	};
	const Row = ({ index, style }) => {
		const move = moves[index]
		const moveName = move.move.name;
		const currentVersion = move.version_group_details.filter((version) => version.version_group.name === currentSelection);
		const learnedAt = currentVersion[0]?.level_learned_at ?? "--";
		const learnedIn = currentVersion[0]?.move_learn_method.name || "--"
		return <GridRow style={style}> 
			<Name>
				{capitalizeFirstLetter(moveName)}
			</Name>
			<LearnedAt>
				{learnedAt}
			</LearnedAt>
			<LearnedIn>
				{capitalizeFirstLetter(learnedIn)}
			</LearnedIn>
		</GridRow>
	};

	return (
		<Wrapper className={className}>
			<FormControl style={{width: "100%", marginBottom: "20px"}}>
				<InputLabel>Pokemon games</InputLabel>
				<Select
					autoWidth={true}
					labelId="pokemon-moves"
					value={currentSelection || ""}
					onChange={handleChange}
				>
					{availableGames.map((game) => {
						return <MenuItem key={game} value={game}>{capitalizeFirstLetter(game)}</MenuItem>
					})}
				</Select>
			</FormControl>
			<GridRow style={{marginBottom: "15px", marginRight: "20px", fontWeight: "bold"}}>
				<Name>
					Name
				</Name>
				<LearnedAt>
					Learned At
				</LearnedAt>
				<LearnedIn>
					Learned In
				</LearnedIn>
			</GridRow>
			<AutoSizer style={{width: "auto", height: "auto"}}>
				{({ width }) => (
					<List
						height={367}
						itemCount={moves.length}
						itemSize={35}
						width={width}
					>
						{Row}
					</List>
				)}
			</AutoSizer>
		</Wrapper>
	)

}

const GridRow = styled.div`
	display: grid;
	grid-template-columns: repeat(3, 1fr);
	grid-template-areas: "name learnedAt learnedIn";
`

const Name = styled.span`
	grid-area: name;
`

const LearnedAt = styled.span`
	grid-area: learnedAt;
`

const LearnedIn = styled.span`
	grid-area: learnedIn;
`

const Wrapper = styled.main`
	width: 100%;
	height: 100%;
	padding: 25px;
	position: relative;
`

const ListItem = styled.span`
`

export { PokemonMovesInfo }