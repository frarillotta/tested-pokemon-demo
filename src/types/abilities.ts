import type { URLString } from './pokemon';

export type Abilities = {
    effect_changes: [],
    effect_entries: {
          effect: string,
          language:{
             name: string,
             url:URLString
          },
          short_effect:string
    }[],
    flavor_text_entries: {
          flavor_text: string,
          language:{
             name: string,
             url: URLString
          },
          version_group:{
            name: string,
            url: URLString
          }
    }[],
    generation:{
       name: string,
       url: URLString
    },
    id: number,
    is_main_series: boolean,
    name: string,
    names:{
          language:{
            name: string,
            url: URLString
          },
          name: string
       }[],
    pokemon: {
          is_hidden: boolean,
          pokemon:{
            name: string,
            url: URLString
          },
          slot: number
    }[]
 }