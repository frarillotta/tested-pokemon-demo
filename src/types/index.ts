export type { pokemonType, pokemonResults } from './pokemonsList'; 
export type { Pokemon, GenerationSprites, URLString, Stat, Move } from "./pokemon"