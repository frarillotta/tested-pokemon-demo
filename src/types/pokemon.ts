import { GenerationSprites } from './pokemon';
export type Pokemon = {
    abilities: {
          ability:{
             name: string,
             url: URLString
          },
          is_hidden: boolean,
          slot: number
       }[],
    base_experience: number,
    forms: {
          name: string,
          url:URLString
       }[],
    game_indices: {
          game_index: number,
          version:{
             name: string,
             url: URLString
          }
    }[],
    height: number,
    held_items: any[],
    id: number,
    is_default: boolean,
    location_area_encounters: string,
    moves: Move[],
    name: string,
    order: number,
    past_types: [],
    species:{
       name: string,
       url: string
    },
    sprites:{
		back_default: URLString | null,
		back_female: URLString | null,
		back_shiny: URLString | null,
		back_shiny_female: URLString | null,
		front_default: URLString | null,
		front_female: URLString | null,
		front_shiny: URLString | null,
		front_shiny_female: URLString | null,
       other:{
          dream_world: GenerationSprites,
          "official-artwork": GenerationSprites
       },
       versions:{
          "generation-i": {
             "red-blue": GenerationSprites,
             yellow: GenerationSprites
          },
            "generation-ii": {
             crystal: GenerationSprites,
             gold: GenerationSprites,
             silver: GenerationSprites
          },
          "generation-iii":{
             emerald: GenerationSprites,
             "firered-leafgreen": GenerationSprites,
             "ruby-sapphire": GenerationSprites
          },
          "generation-iv":{
             "diamond-pearl": GenerationSprites,
             "heartgold-soulsilver": GenerationSprites,
             platinum: GenerationSprites
          },
          "generation-v":{
             "black-white":{
                animated: GenerationSprites,
                GenerationSprites
             }
          },
          "generation-vi":{
             "omegaruby-alphasapphire": GenerationSprites,
             "x-y": GenerationSprites
          },
          "generation-vii":{
             icons: GenerationSprites,
             "ultra-sun-ultra-moon": GenerationSprites
          },
          "generation-viii":{
             icons: GenerationSprites
          }
       }
    },
    stats: Stat[],
    types: {
          slot: number,
          type:{
             name: "normal" | "fire" | "water" | "grass" | "electric" | "ice" | "fighting" | "poison" | "ground" | "flying" | "psychic" | "bug" | "rock" | "ghost" | "dark" | "dragon" | "steel" | "fairy",
             url: URLString
          }
       }[],
    weight: number
 }

export type GenerationSprites = {
    back_default: URLString | null,
    back_female: URLString | null,
    back_shiny: URLString | null,
    back_shiny_female: URLString | null,
    front_default: URLString | null,
    front_female: URLString | null,
    front_shiny: URLString | null,
    front_shiny_female: URLString | null
 };

export type Stat = {
	base_stat: number,
	effort: number,
	stat:{
	   name: "hp" | "attack" | "defense" | "special-defense" | "special-attack" | "speed",
	   url: URLString
	}
};

export type Move = {
	move:{
	   name: string,
	   url: URLString
	},
	version_group_details: {
		  level_learned_at: number,
		  move_learn_method:{
			 name: string,
			 url: URLString
		  },
		  version_group:{
			 name: string,
			 url: URLString
		  }
	  }[],
 }

export type URLString = `https://${string}`;
