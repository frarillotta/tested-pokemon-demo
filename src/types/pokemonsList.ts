export type pokemonType = {
    count: number,
    next: string,
    previous: string,
    results: pokemonResults[]
}

export type pokemonResults = {
    name: string, 
    url: string
}