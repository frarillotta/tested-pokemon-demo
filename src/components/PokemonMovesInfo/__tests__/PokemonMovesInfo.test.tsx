import { render } from "@testing-library/react";
import {PokemonMovesInfo} from "../PokemonMovesInfo";
import moves from "../../../../__mocks__/dataMocks/moves.json";
import { Move } from "../../../types";

describe("PokemonMovesInfo", () => {
	
	describe("when loaded", () => {
		
		it("should have a consistent snapshot", () => {
			const {container} = render(<PokemonMovesInfo moves={moves as Move[]} />);
			expect(container).toMatchSnapshot();
		});

	});

})