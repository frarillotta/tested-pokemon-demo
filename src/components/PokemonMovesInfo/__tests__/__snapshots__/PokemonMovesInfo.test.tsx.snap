// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonMovesInfo when loaded should have a consistent snapshot 1`] = `
<div>
  <main
    class="PokemonMovesInfo__Wrapper-sc-7pl66i-4 fgtUjm"
  >
    <div
      class="MuiFormControl-root"
      style="width: 100%; margin-bottom: 20px;"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
        data-shrink="true"
      >
        Pokemon games
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="pokemon-moves"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          role="button"
          tabindex="0"
        >
          Red-blue
        </div>
        <input
          aria-hidden="true"
          class="MuiSelect-nativeInput"
          tabindex="-1"
          value="red-blue"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="PokemonMovesInfo__GridRow-sc-7pl66i-0 loQwdy"
      style="margin-bottom: 15px; margin-right: 20px; font-weight: bold;"
    >
      <span
        class="PokemonMovesInfo__Name-sc-7pl66i-1 tcijR"
      >
        Name
      </span>
      <span
        class="PokemonMovesInfo__LearnedAt-sc-7pl66i-2 jxpttq"
      >
        Learned At
      </span>
      <span
        class="PokemonMovesInfo__LearnedIn-sc-7pl66i-3 hcKxXz"
      >
        Learned In
      </span>
    </div>
    <div
      style="overflow: visible; height: auto; width: auto;"
    >
      <div
        style="position: relative; height: 367px; width: -50px; overflow: auto; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 70px; width: 100%;"
        >
          <div
            class="PokemonMovesInfo__GridRow-sc-7pl66i-0 loQwdy"
            style="position: absolute; left: 0px; top: 0px; height: 35px; width: 100%;"
          >
            <span
              class="PokemonMovesInfo__Name-sc-7pl66i-1 tcijR"
            >
              Mega-punch
            </span>
            <span
              class="PokemonMovesInfo__LearnedAt-sc-7pl66i-2 jxpttq"
            >
              0
            </span>
            <span
              class="PokemonMovesInfo__LearnedIn-sc-7pl66i-3 hcKxXz"
            >
              Machine
            </span>
          </div>
          <div
            class="PokemonMovesInfo__GridRow-sc-7pl66i-0 loQwdy"
            style="position: absolute; left: 0px; top: 35px; height: 35px; width: 100%;"
          >
            <span
              class="PokemonMovesInfo__Name-sc-7pl66i-1 tcijR"
            >
              Ice-punch
            </span>
            <span
              class="PokemonMovesInfo__LearnedAt-sc-7pl66i-2 jxpttq"
            >
              --
            </span>
            <span
              class="PokemonMovesInfo__LearnedIn-sc-7pl66i-3 hcKxXz"
            >
              --
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1px; height: 1px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </main>
</div>
`;
